includes/blocks/class-llms-blocks-course-progress-block.php

```diff
@@ -0,0 +1,74 @@
+<?php
+/**
+ * Course progress bar block
+ *
+ * @package  LifterLMS_Blocks/Blocks
+ *
+ * @since [version]
+ * @version [version]
+ *
+ * @render_hook llms_course-progress_block_render
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Course progress block class.
+ */
+class LLMS_Blocks_Course_Progress_Block extends LLMS_Blocks_Abstract_Block {
+
+	/**
+	 * Block ID.
+	 *
+	 * @var string
+	 */
+	protected $id = 'course-progress';
+
+	/**
+	 * Is block dynamic (rendered in PHP).
+	 *
+	 * @var bool
+	 */
+	protected $is_dynamic = true;
+
+	/**
+	 * Add actions attached to the render function action.
+	 *
+	 * @since [version]
+	 *
+	 * @param array  $attributes Optional. Block attributes. Default empty array.
+	 * @param string $content    Optional. Block content. Default empty string.
+	 * @return void
+	 */
+	public function add_hooks( $attributes = array(), $content = '' ) {
+
+		add_action( $this->get_render_hook(), array( $this, 'output' ), 10 );
+
+	}
+
+	/**
+	 * Output the course progress bar
+	 *
+	 * @since [version]
+	 *
+	 * @param array $attributes Optional. Block attributes. Default empty array.
+	 * @return void
+	 */
+	public function output( $attributes = array() ) {
+		$progress = do_shortcode( '[lifterlms_course_progress check_enrollment=1]' );
+		$class    = empty( $attributes['className'] ) ? '' : $attributes['className'];
+
+		if ( $progress ) {
+			printf(
+				'<div class="wp-block-llms-%1$s%2$s">%3$s</div>',
+				$this->id,
+				// Take into account the custom class attribute.
+				empty( $attributes['className'] ) ? '' : ' ' . esc_attr( $attributes['className'] ),
+				$progress
+			);
```

Comment:
A few notes on this method....

1) We need a way to override it / customize the output. Most of the existing dynamic blocks output templates and default template override functionality can be used to override their output with PHP. This block will suffer from a lack of extensibility.

I think the easiest way to preserve the extensibility that's currently possible via JS (but now in PHP) would be to add a filter around the html. Instead of doing a printf() switch to $progress = sprintf( ... ) and then after the if run it through a filter and then echo it.

2) I think that it might be beneficial to switch this method to a static method so that the action can be easily removed as well.

Thoughts?

Reply:
I see your point. Let me only add that:
I took the output method from the pricing table block, which is not static - shall we make it static too when we'll be able to break the backward compatibility?

Wanna know my thoughts?
Let's revert this to be a js block, I always like the idea to have more control with PHP but that's not the case now.

Reply:
RE: Pricing table -- ah yes, you're right. It probably should be for the same reasons but I guess when I wrote it I didn't think about it... This is a small issue but yes that block should probably have a static method too....

Reply:
Re; making this a jS block:

I actually think the JS blocks are really the future and if we can reduce the amount of dynamic blocks we use that would be best. Especially since everything being done in this block can be done in JS I don't think it's necessary to make this block dynamic.

I think during this current "transitional" phase it's easier for the common WP dev to extend PHP but that's going to change in the future.